### DISCUSSION TITLE
Reverse proxy forwarding with Apache [The missing guide]

### DISCUSSION THREAD

**Discussion by Dungeonseeker** [coder/code-server#2718] [General]
In case anyone still cannot get this working with Apache heres how I managed it.

Code Server Config
1) `sudo systemctl stop code-server@root (change root to required user)`
2) `sudo nano ~/.config/code-server/config.yaml`
3) Change bind-addr to your machines internal IP (192.168.x.x, 10,0,x,x etc) and if needed add a custom port (IP:PORT)
4) Set password
5) Save and quit
6) `sudo systemctl start code-server@root (change root to required user)`

Apache2 Config
1) `cd /etc/apache2/sites-available/`
2) `sudo nano code.conf`
3) Add the following 

`<VirtualHost *:80>`
`  ServerName code.domain.name.here`
`  RewriteEngine On`
`  RewriteCond %{HTTP:Upgrade} =websocket [NC]`
`  RewriteRule /(.*) ws://192.168.1.45:420/$1 [P,L]`
`  RewriteCond %{HTTP:Upgrade} !=websocket [NC]`
`  RewriteRule /(.*) http://192.168.1.45:420/$1 [P,L]`
`  ProxyRequests off`
`  ProxyPass / http://192.168.1.45:420/ nocanon`
`  ProxyPassReverse / http//192.168.1.45:420/`
`</VirtualHost>`

4) Change ServerName to your own domain
5) Change both Rewrite rules and both ProxyPass addresses to the IP:PORT you set in code-servers config file
*Note: If you didn't specify a custom port then uses code-servers default (8080)
*Note2: edit all of these options to your required settings for the rest of this guide, I won't point it out again
6) Save and quit
7) `sudo a2ensite code`
8) `sudo systemctl restart apache2`

Now you should be able to access code from your specified sub domain.

SSL Config
Before proceeding make sure the above steps are all finished and you can access code server
1) `certbot --apache`
2) Choose your virtual server from the list (or hit enter to add all sub domains)
3) Choose 1 for No HTTP forwarding
4) Certbot should have auto generated the sites SSL config and enabled it
5) `cd /etc/apache2/sites-enabled/`
6) `sudo nano code.conf`
7) Replace existing data with the following (or you can delete the file and recreate a new copy)

`<VirtualHost *:80>`
`    ServerName code.your.domain.here`
`    ServerAdmin webmaster@localhost`
`   Redirect permanent / https://code.your.domain.here`
`</VirtualHost>`

8) Save and quit
9) `sudo nano code-le-ssl.conf`
10) It should look like this

`<IfModule mod_ssl.c>`
`<VirtualHost *:443>`
`    ServerName code.your.domain.here`
`    RewriteEngine On`
`    RewriteCond %{HTTP:Upgrade} =websocket [NC]`
`    RewriteRule /(.*) ws://192.168.1.45:420/$1 [P,L]`
`    RewriteCond %{HTTP:Upgrade} !=websocket [NC]`
`    RewriteRule /(.*) http://192.168.1.45:420/$1 [P,L]`
`    ProxyRequests off`
`    ProxyPass / http://192.168.1.45:420/ nocanon`
`    ProxyPassReverse / http//192.168.1.45:420/`
`    SSLCertificateFile /etc/letsencrypt/live/your.domain.here/fullchain.pem`
`    SSLCertificateKeyFile /etc/letsencrypt/live/your.domain.here/privkey.pem`
`    Include /etc/letsencrypt/options-ssl-apache.conf`
`</VirtualHost>`
`</IfModule>`

11) Save and quit
12) `sudo systemctl restart apache2`

Now code server should be ready on `https://code.yourdomain` and `http://code.yourdomain` will forward users to https.

Hope this helps (maybe add this to the install guide?)

**Comment by jsjoeio** [comment]
Thanks for documenting this @Dungeonseeker ! 🎉 If you'd like, you can open a PR and add this to the docs!
